crates/ra_hir_expand/src/name.rs

```diff
@@ -1,25 +1,87 @@
 //! FIXME: write short doc here
 
-use std::fmt;
+use lasso::{Key, Spur, ThreadedRodeo};
+use once_cell::sync::Lazy;
+use std::{
+    cmp::Ordering,
+    fmt,
+    hash::{BuildHasherDefault, Hash, Hasher},
+    num::NonZeroU32,
+};
 
 use ra_syntax::{ast, SmolStr};
 
+pub use crate::name;
+use rustc_hash::FxHasher;
+
+type Interner = ThreadedRodeo<Spur, BuildHasherDefault<FxHasher>>;
+
+static INTERNER: Lazy<Interner> = Lazy::new(|| {
+    let cap = lasso::Capacity::for_strings(KNOWN_NAMES.len());
+    let interner = Interner::with_capacity_and_hasher(cap, BuildHasherDefault::default());
+    for (i, name) in KNOWN_NAMES.iter().enumerate() {
+        // FIXME: Use `get_or_intern_static` once it's available.
+        let spur: Spur = interner.get_or_intern(name);
+        assert_eq!(unsafe { spur.into_usize() }, i);
+    }
+    interner
+});
+
 /// `Name` is a wrapper around string, which is used in hir for both references
 /// and declarations. In theory, names should also carry hygiene info, but we are
 /// not there yet!
-#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
-pub struct Name(Repr);
+#[derive(Clone, PartialEq, Eq)]
+pub struct Name(NonZeroU32);
+
+/// Compares names lexicographically.
+impl PartialOrd for Name {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        let a = self.to_string();
+        let b = other.to_string();
+        a.partial_cmp(&b)
+    }
+}
+
+impl Ord for Name {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let a = self.to_string();
+        let b = other.to_string();
+        a.cmp(&b)
+    }
+}
+
+/// `Hash` impl looks at interned string to avoid dependency on interning order.
+impl Hash for Name {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        match self.repr() {
+            Repr::Interned(spur) => {
+                INTERNER.resolve(&spur).hash(state);
+            }
+            Repr::TupleField(field) => {
+                field.hash(state);
+            }
+        }
+    }
+}
+
+impl fmt::Debug for Name {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self.repr() {
+            Repr::Interned(spur) => write!(f, "Name({:?})", INTERNER.resolve(&spur)),
+            Repr::TupleField(field) => write!(f, "TupleField({})", field),
+        }
+    }
+}
 
-#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
 enum Repr {
-    Text(SmolStr),
-    TupleField(usize),
+    Interned(Spur),
+    TupleField(u32),
 }
 
 impl fmt::Display for Name {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match &self.0 {
-            Repr::Text(text) => fmt::Display::fmt(&text, f),
+        match self.repr() {
+            Repr::Interned(spur) => fmt::Display::fmt(INTERNER.resolve(&spur), f),
             Repr::TupleField(idx) => fmt::Display::fmt(&idx, f),
         }
     }
```

Comment:
I think we should make INTERNER a field of the database, instead of global. 

That means that we'd have to change every `name.to_string()` to `name.display(db).to_string`, but I think there's no way around that, if we don't want gross hacks in the codebase. 

Seems like a good idea to do that refactoring first (as it would be possible to have `name.display(db)` just delegating to `name.to_string()` and maintain the two together).

I wonder if SSR can help us with automation here.